File: day06/src/main/java/de/havox_design/aoc2017/day06/Day06.java
package de.havox_design.aoc2017.day06;

import de.havox_design.aoc.utils.DataReader;

import java.util.List;

public class Day06 {
    private final List<String> input;

    public Day06(String fileName) {
        input = readData(fileName);
    }

    public static long solvePart1(String fileName) {
        Day06 instance = new Day06(fileName);
        return instance.solvePart1();
    }

    public static long solvePart2(String fileName) {
        Day06 instance = new Day06(fileName);
        return instance.solvePart2();
    }

    public long solvePart1() {
        return 0L;
    }

    public long solvePart2() {return 0L;
    }

    private List<String> readData(String fileName) {
        return DataReader.readData(fileName, MainClass.class);
    }
}

File: day06/src/main/scala/de/havox_design/aoc2017/day06/Day06.scala
package de.havox_design.aoc2017.day06

import scala.annotation.tailrec

object Day06 {
  def solvePart1(filename: String): Int =
    findCycle(parseData(filename))

  def solvePart2(filename: String): Int =
    0

  def main(args: Array[String]): Unit = {
    def dayFileName = "day06.txt"

    println("Solution for part1: " + solvePart1(dayFileName))
    println("Solution for part2: " + solvePart2(dayFileName))
  }

  private def findCycle(data: Seq[Int]): Int =
    calculateCycle(data, Map(), 0)

  @tailrec
  private def calculateCycle(current: Seq[Int], previous: Map[Seq[Int], Int], depth: Int): Int =
    if (previous.contains(current)) {
      depth
    }
    else {
      calculateCycle(calculateStep(current), previous.updated(current, depth), depth + 1)
    }

  private def calculateStep(banks: Seq[Int]): Seq[Int] = {
    val max = banks
      .max
    val start = banks
      .indexWhere(_ == max)
    (1 to max)
      .foldLeft(banks.updated(start, 0)) { (banks, offset) =>
        val index = (start + offset) % banks.size
        banks
          .updated(index, banks(index) + 1)
      }
  }

  private def parseData(filename: String): IndexedSeq[Int] = {
    readData(filename)
      .next()
      .split("\\s+")
      .map(_.trim)
      .toIndexedSeq
      .map(_.toInt)
  }

  private def readData(filename: String): Iterator[String] =
    scala
      .io
      .Source
      .fromResource(filename)
      .getLines()
}
